File: firmware/experiment-bench/portable/telemetry/histogram.c
// SPDX-FileCopyrightText: 2026 Companion Platform contributors
//
// SPDX-License-Identifier: Apache-2.0

#include "telemetry/histogram.h"

#include <stddef.h>

#include "util/sat.h"

int bench_histogram_init(bench_histogram_t *h, const bench_ticks_t *edges,
                         uint32_t bin_count, uint32_t *counts, uint32_t version) {
  h->edges = NULL;
  h->bin_count = 0;
  h->counts = NULL;
  h->underflow = 0;
  h->overflow = 0;
  h->sample_count = 0;
  h->version = version;

  if (edges == NULL || counts == NULL) {
    return 0;
  }
  if (bin_count == 0 || bin_count > BENCH_HISTOGRAM_MAX_BINS) {
    return 0;
  }
  // Bornes STRICTEMENT croissantes : sinon l'appartenance a une classe serait
  // ambigue et l'histogramme silencieusement faux.
  for (uint32_t i = 0; i < bin_count; i++) {
    if (!(edges[i] < edges[i + 1])) {
      return 0;
    }
  }
  for (uint32_t i = 0; i < bin_count; i++) {
    counts[i] = 0;
  }
  h->edges = edges;
  h->bin_count = bin_count;
  h->counts = counts;
  return 1;
}

void bench_histogram_add(bench_histogram_t *h, bench_ticks_t value) {
  if (h->edges == NULL || h->counts == NULL || h->bin_count == 0) {
    return;  // histogramme non initialise : ne fabrique aucun chiffre
  }
  bench_sat_inc_u32(&h->sample_count);

  if (value < h->edges[0]) {
    bench_sat_inc_u32(&h->underflow);
    return;
  }
  if (value >= h->edges[h->bin_count]) {
    bench_sat_inc_u32(&h->overflow);
    return;
  }
  // Classe [edges[i], edges[i+1]) : borne basse incluse, borne haute exclue.
  for (uint32_t i = 0; i < h->bin_count; i++) {
    if (value >= h->edges[i] && value < h->edges[i + 1]) {
      bench_sat_inc_u32(&h->counts[i]);
      return;
    }
  }
}

File: tools/measurement/measurement/analysis/histogram.py
# SPDX-FileCopyrightText: 2026 Companion Platform contributors
#
# SPDX-License-Identifier: Apache-2.0
"""Histogramme de latence : VUE DERIVEE, jamais autoritaire.

L'histogramme recalcule ICI, depuis la serie brute, fait foi. Celui eventuellement
produit par le MCU est SECONDAIRE : il est compare a ce recalcul, et toute
divergence est un defaut d'instrumentation ou d'implementation a investiguer,
pas un resultat a arbitrer.

Convention de classe, identique au firmware : ``[edges[i], edges[i+1])`` — borne
basse INCLUSE, borne haute EXCLUE."""

from typing import Any


def derive(values: list[int], edges: list[int]) -> dict[str, Any]:
    """Construit l'histogramme d'une population depuis la serie BRUTE."""
    if len(edges) < 2:
        raise ValueError("il faut au moins deux bornes")
    for i in range(len(edges) - 1):
        if not edges[i] < edges[i + 1]:
            raise ValueError("bornes non strictement croissantes")

    bin_count = len(edges) - 1
    counts = [0] * bin_count
    underflow = 0
    overflow = 0
    for v in values:
        if v < edges[0]:
            underflow += 1
        elif v >= edges[-1]:
            overflow += 1
        else:
            for i in range(bin_count):
                if edges[i] <= v < edges[i + 1]:
                    counts[i] += 1
                    break
    return {
        "bin_edges": list(edges),
        "bin_counts": counts,
        "underflow": underflow,
        "overflow": overflow,
        "sample_count": len(values),
    }


def compare(device: dict[str, Any] | None, tooling: dict[str, Any]) -> dict[str, Any]:
    """Compare l'histogramme embarque au recalcul autoritaire.

    ``device`` a None signifie que l'histogramme embarque etait desactive (cas
    par DEFAUT) : ce n'est PAS une divergence."""
    if device is None:
        return {"compared": False, "reason": "histogramme embarque desactive"}

    differences: list[str] = []
    if list(device.get("bin_edges", [])) != list(tooling["bin_edges"]):
        differences.append("bin_edges")
    for key in ("bin_counts", "underflow", "overflow", "sample_count"):
        if device.get(key) != tooling[key]:
            differences.append(key)
    return {
        "compared": True,
        "match": not differences,
        "differences": differences,
    }
